### PR TITLE
Discorick/stripe webhook

### PR DESCRIPTION
**Pull request by discorick**
Properly implements the stripe webhook and tests it

### CODE REVIEW

File: app/routes/api/webhooks.rb [hunk 1]
```diff
@@ -39,19 +39,25 @@ class Webhooks < Base
           json message: "Webhook received"
         end
 
-        post '/api/stripe/webhook' do
-          payload = Hashie::Mash.new JSON.parse(params[:payload])
-          id = payload.data.object.id
+        #Thinking Ahead to API 2.0, all this logic should be encapsulated into an
+        #event handler / background job
+        post '/api/site/stripe/webhook' do
+          return json message: "Not Authorized" if params[:token] != ENV["STRIPE_WEBHOOK_TOKEN"]
```
Comment (emadrid) [inline app/routes/api/webhooks.rb]: `unless` reads better than `if` here

File: app/routes/api/webhooks.rb [hunk 2]
```diff
@@ -39,19 +39,25 @@ class Webhooks < Base
...
+
+          payload = Hashie::Mash.new(params)
+          id = payload.data.object.customer
```
Comment (emadrid) [inline app/routes/api/webhooks.rb]: does `Hashie::Mash` ensure that these `payload` child objects are not nil?

File: app/routes/api/webhooks.rb [hunk 3]
```diff
@@ -39,19 +39,25 @@ class Webhooks < Base
...
 
-          query = Queries::CouchCustomer.get(id, couch)
+          query = Queries::CouchCustomer.get_cust(id, couch)
           doc = QueryHandler.exec(&query)
 
-          if payload.type == "customer.subscription.updated"
-            plan_doc = docs.rows.first.value
-            plan_doc.stripe.customer = payload.data.object
+          if doc[:rows] && payload.type == "customer.subscription.updated"
```
Comment (emadrid) [inline app/routes/api/webhooks.rb]: we can check `payload.type` before executing the query

File: app/routes/api/webhooks.rb [hunk 4]
```diff
@@ -39,19 +39,25 @@ class Webhooks < Base
...
+            plan_doc = doc.rows.first.value
             plan_doc.trial = "expired" if payload.data.object.status != "trialing"
```
Comment (emadrid) [inline app/routes/api/webhooks.rb]: we can set `plan_doc.trial` to expired without querying couch

File: app/routes/api/webhooks.rb (hunk 2 above)
Comment (discorick) [inline app/routes/api/webhooks.rb]: Line 45 will more or less verify the payload is there.

File: app/routes/api/webhooks.rb (hunk 3 above)
Comment (discorick) [inline app/routes/api/webhooks.rb]: Good Call.

File: spec/lib/webhooks/stripe/stripe_spec.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
+require "spec_helper"
+#requires that the webapp is running on port 5000
+
+describe "Stripe Webhooks", :webhooks do
+
+  let(:api_path) { "http://localhost:5000/api/site/stripe/webhook"}
+  let(:stripe_token) { ENV["STRIPE_WEBHOOK_TOKEN"] }
```
Comment (emadrid) [inline spec/lib/webhooks/stripe/stripe_spec.rb]: does the toke have to come from an env var?

File: app/routes/api/webhooks.rb (hunk 4 above)
Comment (discorick) [inline app/routes/api/webhooks.rb]: No way to update a couch doc without querying it first that I know of.

File: spec/lib/webhooks/stripe/stripe_spec.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  end
+
+  after(:each) do
+    up_to_date_doc = @couch.get(@doc["id"])
+    @couch.delete_doc(up_to_date_doc)
+  end
+
+  describe "customer.subscription.updated" do
+
+    let(:stripe_data){ File.read(@fixtures_path + "stripe_subscription_update.json") }
```
Comment (emadrid) [inline spec/lib/webhooks/stripe/stripe_spec.rb]: use File.join to concat the fixtures_path and json

File: spec/lib/webhooks/stripe/stripe_spec.rb (hunk 5 above)
Comment (discorick) [inline spec/lib/webhooks/stripe/stripe_spec.rb]: Stripe doesn't provide any mechanism whatsoever to auth webhooks, so its either that or we stash a key in couch.

File: spec/lib/webhooks/stripe/stripe_spec.rb [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  before(:each) do
+    @doc = seed("couch_trial_account_seed.json")
+  end
+
+  after(:each) do
+    up_to_date_doc = @couch.get(@doc["id"])
+    @couch.delete_doc(up_to_date_doc)
+  end
+
+  describe "customer.subscription.updated" do
```
Comment (emadrid) [inline spec/lib/webhooks/stripe/stripe_spec.rb]: do we have a spec for a "trial" customer?

File: spec/spec_helper.rb [hunk 8]
```diff
@@ -1 +1,26 @@
...
+  test_customer = JSON.parse(File.read(@fixtures_path + fixture))
+  doc = @couch.save_doc(test_customer)
+  doc
+end
+
+RSpec.configure do |c|
+  c.filter_run_excluding :webhooks
+
+  c.around do |example|
+    if example.metadata[:webhooks]
```
Comment (emadrid) [inline spec/spec_helper.rb]: :+1:

File: app/routes/api/webhooks.rb (hunk 4 above)
Comment (rauhryan) [inline app/routes/api/webhooks.rb]: true, you need the rev # to update documents

File: spec/lib/webhooks/stripe/stripe_spec.rb (hunk 5 above)
Comment (rauhryan) [inline spec/lib/webhooks/stripe/stripe_spec.rb]: +1 for env

File: spec/lib/webhooks/stripe/stripe_spec.rb (hunk 6 above)
Comment (discorick) [inline spec/lib/webhooks/stripe/stripe_spec.rb]: If performance is the issue then interpolation is far faster than File.join

**Pull request closed without merging**